### DISCUSSION TITLE
Adding affiliate marketing

### DISCUSSION THREAD

**Discussion by adamekthe** [medusajs/medusa#4684] [Feature Requests]
The Idea
I think it would be a great idea to add affiliate marketing system to Medusa. So we can leverage this for better store performance and more sales. Signed-in users would be able to share their generated medusa link for other customers and would be able to get parts of the profit from item sales.

Flow
-User creates an account and generates unique affiliate link for example [www.mystore.com/affiliate/UniqueID](http://www.mystore.com/affiliate/UniqueID)
-When other user enters the link a cookie is placed storing the ID
-If user makes a purchase a % of the sale is awarded to the affiliate account
-Affiliates could be tracked and managed in Admin panel

Possible problems
Sending funds to the affiliates could be problematic. But it could be resolved with stripe integration or manual payments.

**Comment by LuigiClemente** [comment]
```mermaid
sequenceDiagram
    participant User
    participant Application
    participant Commission

    User->>Application: Makes a purchase using a discount
    Application->>DiscountRule: Applies discount
    DiscountRule->>DiscountCondition: Checks conditions
    DiscountRule->>Commission: Calculate commission
    DiscountRule-->>Application: Apply discount and commission
    Commission-->>Application: Store commission
    Application-->>User: Complete purchase with discount and commission
```

# Extending Medusa JS for Affiliate Program with Commission

## Introduction

In this documentation, we will discuss how to extend the Medusa JS open-source application to support an affiliate program with commission functionality. The goal is to enable discounts to be linked to users, where these users can earn commissions when a sale is made using the applied discount. This feature enhancement requires adapting the existing structure, primarily focusing on entities such as DiscountRule and DiscountCondition.

## Feature Overview

The extended functionality aims to allow users to earn commissions through discounts, with the following key features:

* Discounts can be linked to specific users.
* Commissions can be set up using parameters similar to discounts, either as a fixed amount or a percentage of the sale.

## Adaptation Steps

To achieve this, follow these adaptation steps:

1. **Introduce the Commission Entity:** Create a new entity called "Commission" to track and store commissions earned by users. The Commission entity should include:
    
    * `user_id`: Identifying the user earning the commission.
    * `amount`: Storing the commission amount.
2. **Modify the DiscountRule Entity:** Enhance the DiscountRule entity to support user-specific discounts and commission parameters. Add the following attributes:
    
    * `user_id`: Links the discount to a specific user eligible for commission.
    * `commission_percentage`: Specifies the percentage of the sale amount that the user will earn as commission.
    * `fixed_commission_amount`: Allows specifying a fixed amount of commission.
3. **Modify the DiscountCondition Entity:** Update the DiscountCondition entity to indicate conditions associated with earning commissions. Introduce an attribute:
    
    * `earn_commission`: A boolean attribute that defines whether a condition is linked to commission earnings.
4. **Revise the Discount Application Process:** Adapt the discount application process to account for commission calculations:
    
    * Calculate the commission amount based on the `commission_percentage` or `fixed_commission_amount` specified in the DiscountRule.
    * Create a Commission record with the `user_id` and the commission amount.

## Medusa required modules

* [Discounts Module](https://docs.medusajs.com/modules/discounts)
* [Users Module](https://docs.medusajs.com/modules/users)

**Comment by LuigiClemente** [comment]
import { Entity, Column, Index, ManyToOne, JoinColumn, BeforeInsert, OneToMany } from "typeorm";
import { SoftDeletableEntity } from "../interfaces/models/soft-deletable-entity";
import { DiscountRule } from "./discount-rule";
import { DiscountCondition } from "./discount-condition";
import { User } from "./user";
import { Commission } from "./commission";
import { Order } from "./order";

@Entity()
export class Discount extends SoftDeletableEntity {
  // Discount Code: A unique identifier for the discount.
  @Index({ unique: true, where: "deleted_at IS NULL" })
  @Column()
  code: string;

  // Dynamic Discount: Indicates if the discount is dynamic.
  @Column()
  is_dynamic: boolean;

  // Related DiscountRule: ID of the associated DiscountRule.
  @Index()
  @Column({ nullable: true })
  rule_id: string;

  // Rule Entity: Reference to the related DiscountRule entity.
  @ManyToOne(() => DiscountRule, { cascade: true })
  @JoinColumn({ name: "rule_id" })
  rule: DiscountRule;

  // Disabled Status: Indicates if the discount is currently disabled.
  @Column()
  is_disabled: boolean;

  // Parent Discount: ID of the parent discount if applicable.
  @Column({ nullable: true })
  parent_discount_id: string;

  // Parent Discount Entity: Reference to the parent discount entity.
  @ManyToOne(() => Discount, { nullable: true })
  @JoinColumn({ name: "parent_discount_id" })
  parent_discount: Discount;

  // Start Date: The date when the discount becomes effective.
  @Column({ type: "timestamptz", default: () => "CURRENT_TIMESTAMP" })
  starts_at: Date;

  // End Date: The date when the discount expires (nullable for ongoing discounts).
  @Column({ type: "timestamptz", nullable: true })
  ends_at: Date | null;

  // Validity Duration: Duration of validity (nullable for open-ended discounts).
  @Column({ nullable: true })
  valid_duration: string | null;

  // Usage Limit: Maximum number of times the discount can be used (nullable for unlimited use).
  @Column({ nullable: true })
  usage_limit: number | null;

  // Usage Count: Current count of how many times the discount has been used.
  @Column({ default: 0 })
  usage_count: number;

  // User Link: Reference to the related User entity for user-specific discounts.
  @ManyToOne(() => User, { cascade: true, nullable: true })
  @JoinColumn({ name: "user_id" })
  user: User;

  // Commission Percentage: Percentage of sale amount earned as commission (nullable).
  @Column({ type: "decimal", precision: 5, scale: 2, nullable: true })
  commission_percentage: number | null; // Allow null for flexibility

  // Fixed Commission Amount: Fixed amount commission (default: 0).
  @Column({ type: "decimal", precision: 10, scale: 2, default: 0 })
  fixed_commission_amount: number;

  // Related Commissions: Commission records associated with this discount.
  @OneToMany(() => Commission, (commission) => commission.discount)
  commissions: Commission[];

  // Related Orders: Orders associated with earned Commissions.
  @OneToMany(() => Order, (order) => order.commission)
  orders: Order[];

  /**
   * BeforeInsert Method: Converts the discount code to uppercase and trims whitespace.
   * This method ensures that discount codes are consistently formatted for storage and retrieval.
   */
  @BeforeInsert()
  private upperCaseCodeAndTrim(): void {
    this.code = this.code.toUpperCase().trim();
  }

  /**
   * Validate commission_percentage or fixed_commission_amount.
   * Ensure only one commission configuration is set.
   */
  private validateCommissionConfig(): boolean {
    return (this.commission_percentage === null && this.fixed_commission_amount !== null) ||
           (this.commission_percentage !== null && this.fixed_commission_amount === null);
  }

  // Additional Functionality:
  // You can potentially add cascading deletes for DiscountRule-related Commissions.
  // Utility methods for commission total calculations on User and DiscountRule can be implemented.
}

**Comment by LuigiClemente** [comment]
# Medusa Commerce Application - Discount Entity Documentation

## Introduction

The Discount entity in the Medusa commerce application represents a discount offered to customers. This document provides detailed information about the Discount entity, including its attributes, relationships, and customization options.

## Entity Overview

The Discount entity is responsible for managing various aspects of discounts, such as code, eligibility, and associated rules. We have made some modifications to accommodate additional functionality related to commission types.

## Attributes

* `code` (string): A unique identifier for the discount.
* `is_dynamic` (boolean): Indicates if the discount is dynamic, allowing it to change over time.
* `rule_id` (string): ID of the associated DiscountRule.
* `is_disabled` (boolean): Indicates if the discount is currently disabled.
* `parent_discount_id` (string): ID of the parent discount if applicable.
* `starts_at` (Date): The date when the discount becomes effective.
* `ends_at` (Date | null): The date when the discount expires (nullable for ongoing discounts).
* `valid_duration` (string | null): Duration of validity (nullable for open-ended discounts).
* `usage_limit` (number | null): Maximum number of times the discount can be used (nullable for unlimited use).
* `usage_count` (number): Current count of how many times the discount has been used.
* `user` (User): Reference to the related User entity for user-specific discounts.
* `commission_percentage` (number | null): Percentage of the sale amount earned as commission (nullable).
* `fixed_commission_amount` (number): Fixed amount commission (default: 0).

## Relationships

* `rule` (DiscountRule): Reference to the related DiscountRule entity.
* `parent_discount` (Discount): Reference to the parent discount entity if applicable.
* `commissions` (Commission[]): Commission records associated with this discount.
* `orders` (Order[]): Orders associated with earned Commissions.

## BeforeInsert Method

The BeforeInsert method ensures that discount codes are consistently formatted for storage and retrieval. It converts the discount code to uppercase and trims whitespace.

## Commission Configuration

To accommodate commission-related changes, we have introduced two commission configuration attributes:

* `commission_percentage` (number | null): Percentage of the sale amount earned as commission (nullable).
* `fixed_commission_amount` (number): Fixed amount commission (default: 0).

The `validateCommissionConfig` method ensures that only one commission configuration is set, either `commission_percentage` or `fixed_commission_amount`.

## Additional Functionality

You can enhance the Discount entity by adding utility methods for total calculations related to commission types, cascading deletes for DiscountRule-related Commissions, or any other custom functionality as needed.

## Migration

To add the fixed commission columns and reflect the changes in the database, you may need to create a migration script. Below is an example of a migration script that adds these columns:

```javascript
// src/migrations/20230306000300-add-commission-columns-to-discount.js

import { MigrationInterface, QueryRunner } from "typeorm";

export class AddCommissionColumnsToDiscount20230306000300 implements MigrationInterface {

  public async up(queryRunner: QueryRunner): Promise<void> {

    // Add user_id column
    await queryRunner.query(`
      ALTER TABLE "discount" 
      ADD COLUMN "user_id" uuid NULL
    `);

    // Add foreign key constraint 
    await queryRunner.query(`
      ALTER TABLE "discount"
      ADD CONSTRAINT "fk_discount_user"  
      FOREIGN KEY ("user_id")
      REFERENCES "user"("id")
      ON DELETE CASCADE
    `);

    // Add commission_percentage column
    await queryRunner.query(`
      ALTER TABLE "discount"
      ADD COLUMN "commission_percentage" numeric(5,2) NULL 
    `);

    // Add fixed_commission_amount column
    await queryRunner.query(`
      ALTER TABLE "discount"
      ADD COLUMN "fixed_commission_amount" numeric(10,2) DEFAULT 0 NOT NULL
    `);
  }

  public async down(queryRunner: QueryRunner): Promise<void> {

    // Drop commission_percentage column
    await queryRunner.query(`
      ALTER TABLE "discount"
      DROP COLUMN "commission_percentage"
    `);

    // Drop fixed_commission_amount column
    await queryRunner.query(`
      ALTER TABLE "discount"
      DROP COLUMN "fixed_commission_amount"
    `);

    // Drop foreign key constraint
    await queryRunner.query(`
      ALTER TABLE "discount"
      DROP CONSTRAINT "fk_discount_user"
    `);

    // Drop user_id column
    await queryRunner.query(`
      ALTER TABLE "discount"
      DROP COLUMN "user_id"
    `);
  }
}

```

## TypeScript Declaration File

A TypeScript declaration file is used to make TypeScript aware of the custom attributes. Here's an example declaration file:

```typescript
// src/index.d.ts

declare module "@medusajs/medusa" {
  interface Discount {
    user: User;
    commission_percentage: number;
    fixed_commission_amount: number;
  }
}
```

## Model Extension

To create an extended Discount entity with the changes, you can create a model file as follows:

```typescript
// src/models/discount.ts

import { Entity } from "typeorm";
import { Discount as BaseDiscount } from "@medusajs/medusa";

@Entity()
export class Discount extends BaseDiscount {
  // Custom attributes and relations can be added here.
}
```

## Loader File

A loader file allows custom attributes to be returned from API routes. Here's an example loader file:

```javascript
// src/loaders/discount-fields.ts

export default async function () {
  const imports = await import("@medusajs/medusa/dist/api/routes/store/discounts");
  imports.allowedDiscountFields = [
    ...imports.allowedDiscountFields,
    "user",
    "commission_percentage",
    "fixed_commission_amount"
  ];
}
```

## Building

You can compile the TypeScript files by adding a build script to your project:

```json
{
  "scripts": {
    "build": "tsc"
  }
}
```

This documentation covers the Discount entity in the Medusa commerce application, along with the changes related to commission types and customization options for specific needs.

**Comment by LuigiClemente** [comment]
## Extending the Medusa JS Admin UI to Add a Discount Extension Menu

This guide will illustrate the logic of extending the Medusa JS Admin UI to add a discount extension menu. The extension menu will allow users to create discounts that are applied to a specific user.

**Prerequisites:**

-   Node.js v16 or higher
-   Yarn v1.22 or higher
-   Medusa JS v0.13 or higher

**Steps:**

1.  Create a new directory for your extension and navigate to it.
2.  Initialize a new Node.js project.
3.  Install the Medusa JS admin extension dependencies.

4.  Create a new file called `src/admin/index.js` and add the following code:

```
import { AdminExtension } from '@medusajs/admin';

class DiscountExtension extends AdminExtension {
  constructor() {
    super();

    // Add a new menu item to the Discounts page.
    this.addMenuItem('Discounts', {
      label: 'Discount Extensions',
      icon: 'Gift',
      route: '/discounts/extensions',
    });

    // Register a new route for the discount extension menu.
    this.addRoute('/discounts/extensions', {
      component: () => import('./DiscountExtensionMenu'),
    });
  }
}

export default DiscountExtension;
```

5.  Create a new file called `src/admin/DiscountExtensionMenu.js` and add the following code:

```
import React from 'react';
import { Form, Input, Radio, Button } from '@medusajs/admin';

class DiscountExtensionMenu extends React.Component {
  state = {
    type: 'percentage',
    amount: '',
    user: null,
  };

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = async () => {
    // Create a new discount object.
    const discount = {
      type: this.state.type,
      amount: this.state.amount,
      users: [this.state.user],
    };

    // Create the discount using the Medusa JS Admin API.
    const response = await fetch('/admin/discounts', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(discount),
    });

    // Check the response status code.
    if (response.status === 201) {
      // The discount was created successfully.
      // Redirect the user to the discounts page.
      window.location.href = '/admin/discounts';
    } else {
      // An error occurred while creating the discount.
      // Display an error message to the user.
    }
  };

  render() {
    return (
      <Form onSubmit={this.handleSubmit}>
        <Radio
          name="type"
          value="percentage"
          checked={this.state.type === 'percentage'}
          onChange={this.handleChange}
        >
          Percentage
        </Radio>
        <Radio
          name="type"
          value="fixed_amount"
          checked={this.state.type === 'fixed_amount'}
          onChange={this.handleChange}
        >
          Fixed Amount
        </Radio>
        <Input
          name="amount"
          type={this.state.type === 'percentage' ? 'number' : 'text'}
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.handleChange}
        />
        <Input
          name="user"
          type="text"
          placeholder="User"
          value={this.state.user}
          onChange={this.handleChange}
        />
        <Button type="submit">Create Discount</Button>
      </Form>
    );
  }
}

export default DiscountExtensionMenu;
```

6.  Build your extension.

7.  Update your Medusa JS configuration file to enable your extension.

```
module.exports = {
  plugins: [
    {
      resolve: '@medusajs/extension-discount',
    },
  ],
};
```

8.  Restart the Medusa JS server.

**Comment by radoslaw-sz** [comment]
hi @adamekthe 
I have created small part of affiliate marketing which covers affiliate discounts. Maybe you will find it useful:
https://www.npmjs.com/package/@rsc-labs/medusa-affiliate-discount

**Comment by JR132662** [comment]
Hi @adamekthe message me if youre still searching for this. im building an NPM package that is pretty much one click, you just have to design the look of your affiliate module and set up keys to pay them
